### PR TITLE
Lesson15

### PR DESCRIPTION
**Pull request by Naumych**

### CODE REVIEW

**Comment by svidlonix** [comment]
Please fix conflicts 

Conflicting files:
app/models/carriage.rb
app/models/train.rb
app/views/trains/show.html.erb

File: app/models/carriage.rb [hunk 1]
```diff
@@ -1,6 +1,30 @@
...
+
+  def self.carriage_types
+    %w[EconomyCarriage CoupeCarriage SVCarriage SeatedCarriage]
+  end
+
+  def add_number_in_train
+    carriages_in_train = Carriage.where(train_id: train_id)
+    number_in_train_presented = carriages_in_train.select { |carriage| carriage.number_in_train.present? }
+
+    self.number_in_train = if number_in_train_presented.present?
```
Comment (svidlonix) [inline app/models/carriage.rb]: where do you find this life hook with self?

File: app/models/railway_stations_route.rb [hunk 2]
```diff
@@ -3,4 +3,25 @@
 class RailwayStationsRoute < ApplicationRecord
   belongs_to :railway_station
   belongs_to :route
+
+  validates :railway_station_id, uniqueness: { scope: :route_id }
+  # validates :position, uniqueness: { scope: :route_id }
```
Comment (svidlonix) [inline app/models/railway_stations_route.rb]: Please delete useless code.

File: app/models/railway_stations_route.rb [hunk 3]
```diff
@@ -3,4 +3,25 @@
...
+                      position_present.map(&:position).max + 1
+                    else
+                      1
+                    end
+    save
+  end
+
+  def update_position; end
 end
+# route.railway_stations_routes.where(position: true).
```
Comment (svidlonix) [inline app/models/railway_stations_route.rb]: Please delete the useless code.

File: app/models/route.rb [hunk 4]
```diff
@@ -1,9 +1,31 @@
...
-  has_many :railway_stations_routes
+  has_many :railway_stations_routes, dependent: :destroy
   has_many :railway_stations, through: :railway_stations_routes
   has_many :trains
+
+  validates :name, presence: true
+  validate :stations_count
+
+  before_validation :set_name
+  # after_save :add_position
```
Comment (svidlonix) [inline app/models/route.rb]: Please delete the useless code.

File: app/models/route.rb [hunk 5]
```diff
@@ -1,9 +1,31 @@
...
+
+  def update_position(station, position)
+    exist_station_with_position = RailwayStationsRoute.find_by(route_id: 5, position: position) # существующая станция с позицоей position
```
Comment (svidlonix) [inline app/models/route.rb]: Please don't hardcode values.

File: app/models/train.rb [hunk 6]
```diff
@@ -8,18 +8,24 @@ class Train < ApplicationRecord
...
-
   def train_carriages_info
-    types = carriages.map(&:carriage_type).uniq
-    types.each do |type|
-      carriages_by_type = carriages.select { |carriage| carriage.carriage_type == type }
-      train_carriages[type] = { top_seats_count: carriages_by_type.sum(&:top_seats_count),
-                                lower_seats_count: carriages_by_type.sum(&:lower_seats_count),
-                                size: carriages_by_type.size }
+    train_carriages = {}
+    Carriage.carriage_types.each do |type|
```
Comment (svidlonix) [inline app/models/train.rb]: Looks like this functionality preparing data for view please move this functionality to the helper.

File: app/models/train.rb [hunk 7]
```diff
@@ -8,18 +8,24 @@ class Train < ApplicationRecord
...
+      carriages_by_type = carriages.select { |carriage| carriage.type == type }
+      train_carriages[type] = { count: carriages_by_type.size,
```
Comment (svidlonix) [inline app/models/train.rb]: It looks huge and ugly.

File: app/models/train.rb [hunk 8]
```diff
@@ -8,18 +8,24 @@ class Train < ApplicationRecord
...
+                                top_seats: carriages_by_type.sum(&:top_seats),
+                                side_top_seats: carriages_by_type.sum(&:side_top_seats),
+                                side_bottom_seats: carriages_by_type.sum(&:side_bottom_seats),
+                                seats: carriages_by_type.sum(&:seats) }
     end
     train_carriages
   end
+
+  def ordered
```
Comment (svidlonix) [inline app/models/train.rb]: It looks huge and ugly.

File: app/views/trains/show.html.erb [hunk 9]
```diff
@@ -14,23 +14,25 @@
...
-    <td>lower seats count </td>
-    <td><%= @train.train_carriages['compartment'][:lower_seats_count] %></td>
-  </tr>
+  <% @train_carriages.each do |type, count_values |%>
+    <td>
+      <th><%= "#{type}" %> counts: </th>
+    </td>
+      <% count_values.each do |seat, count| %>
+      <tr>
+            <th><%= "#{seat}:" %></th>
```
Comment (svidlonix) [inline app/views/trains/show.html.erb]: Please fix indent.

File: db/migrate/20190815153504_create_carriages.rb [hunk 10]
```diff
@@ -1,10 +0,0 @@
-class CreateCarriages < ActiveRecord::Migration[5.2]
```
Comment (svidlonix) [inline db/migrate/20190815153504_create_carriages.rb]: It comments doesn't need fixe but It very bad practices delete migration please don't do it in future.

File: app/models/carriage.rb (hunk 1 above)
Comment (Naumych) [reply app/models/carriage.rb]: Is thats wrong? Should I rewrite with update(number_in_train: )?

**Pull request merged**